src/util/shared_exclusive_lock.c

```diff
@@ -0,0 +1,178 @@
+/*
+ * Copyright (c) 2006-Present, Redis Ltd.
+ * All rights reserved.
+ *
+ * Licensed under your choice of the Redis Source Available License 2.0
+ * (RSALv2); or (b) the Server Side Public License v1 (SSPLv1); or (c) the
+ * GNU Affero General Public License v3 (AGPLv3).
+*/
+#include "shared_exclusive_lock.h"
+#include <pthread.h>
+#include <stdatomic.h>
+#include <time.h>
+#include "rmutil/rm_assert.h"
+
+
+// Lock used as an alternative to the GIL, this is an additional lock which is always acquired
+// when calling SharedExclusiveLock_Acquire, and is held until SharedExclusiveLock_Release is called.
+// When the GIL is lent by the main thread, threads will be satisfied by acquiring this lock only.
+pthread_mutex_t GILAlternativeLock = PTHREAD_MUTEX_INITIALIZER;
+// Lock to synchronize internal state.
+// Expects low contention (at most one thread and the main thread will race on it).
+pthread_mutex_t InternalLock = PTHREAD_MUTEX_INITIALIZER;
+// Condition for threads waiting to acquire the GIL while it is not available.
+// It is also used with a timeout because we cannot guarantee that the main thread will always signal it.
+// GILAlternativeLock must be held when waiting on this condition.
+pthread_cond_t GILAvailable = PTHREAD_COND_INITIALIZER;
+// Condition for the main thread to wait on while trying to take back the GIL.
+pthread_cond_t GILIsBorrowed = PTHREAD_COND_INITIALIZER;
+// Flags to indicate whether the GIL is lent by the main thread
+bool GIL_lent = false;
+// Flag to indicate whether the GIL is currently borrowed by any thread
+bool GIL_borrowed = false;
+
+// Locks Order:
+// 1. GIL
+// 2. GILAlternativeLock
+// 3. InternalLock
+// A thread may hold any combination of these locks, but must always acquire them in this order.
+// One exception is that we acquire the GIL last in SharedExclusiveLock_Acquire, but there is
+// no risk of deadlock because we do it with a "try" mechanism.
+
+#define TIMEOUT_NANOSECONDS 5000 // 5 us in nanoseconds
+#define NANOSEC_PER_SECOND 1000000000L // 10^9
+
+
+static inline void set_timeout(struct timespec *ts) {
+  #ifdef __APPLE__
+      ts->tv_sec  = 0; // (time_t)(TIMEOUT_NANOSECONDS / NANOSEC_PER_SECOND);
+      ts->tv_nsec = TIMEOUT_NANOSECONDS; // (long)(TIMEOUT_NANOSECONDS % NANOSEC_PER_SECOND);
+  #else
+      struct timespec now;
+      clock_gettime(CLOCK_MONOTONIC_RAW, &now);  // NOT *_RAW
+
+      // Add TIMEOUT_NANOSECONDS to 'now' with proper carry
+      uint64_t add_ns = (uint64_t)TIMEOUT_NANOSECONDS;
+      uint64_t ns = (uint64_t)now.tv_nsec + (add_ns % NANOSEC_PER_SECOND);
+      ts->tv_sec  = now.tv_sec + (time_t)(add_ns / NANOSEC_PER_SECOND) + (time_t)(ns / NANOSEC_PER_SECOND);
+      ts->tv_nsec = (long)(ns % NANOSEC_PER_SECOND);
```

Comment:
Consider avoiding division and modulus operations, as `TIMEOUT_NANOSECONDS <= NANOSEC_PER_SECOND` and at most we will overflow by 1 second (which fits in `tv_nsec`).
We can also ensure that with a static assert.